plasmapy/classes/sources/openpmd_hdf5.py

```diff
@@ -0,0 +1,136 @@
+import h5py
+import numpy as np
+import astropy.units as u
+
+from plasmapy.classes import GenericPlasma
+from plasmapy.utils import OpenPMDError
+
+import os
+from distutils.version import StrictVersion
+
+
+_OUTDATED_VERSION = "1.1.0"
+_NEWER_VERSION = "2.0.0"
+
+# This is the order what OpenPMD uses to store unit
+# dimensions for a record.
+_UNITS = (u.meter,
+          u.kilogram,
+          u.second,
+          u.ampere,
+          u.Kelvin,
+          u.mol,
+          u.candela)
+
+
+def _fetch_units(openPMD_dims):
+    """
+    Converts a collection of OpenPMD dimensions to astropy.units.
+    """
+
+    units = u.dimensionless_unscaled
+    for factor, unit in zip(openPMD_dims, _UNITS):
+        units *= (unit ** factor)
+    units, *_ = units.compose()
+    return units
+
+
+def _valid_version(openPMD_version,
+                   outdated=_OUTDATED_VERSION, newer=_NEWER_VERSION):
+    """
+    Checks if the passed version is supported or not.
+    """
+
+    parsed_version = StrictVersion(openPMD_version)
+    outdated_version = StrictVersion(outdated)
+    newer_version = StrictVersion(newer)
+    return outdated_version <= parsed_version < newer_version
+
+
+class HDF5Reader(GenericPlasma):
+    def __init__(self, hdf5, **kwargs):
+        """
+        Core class for accessing various attributes on HDF5 files that
+        are based on OpenPMD standards.
+
+        Attributes
+        ----------
+        electric_field : `astropy.units.Quantity`
+            An (x, y, z) array containing electric field data.
+        charge_density : `astropy.units.Quantity`
+            An array containing charge density data.
```

Comment:
So... my bad here, but I guess I kind of forgot that the 2D  and 3D examples were showing off electrostatic + particle simulations. If you look at the thetaMode example, you will find two more mesh quantities:

```
(base) 20:04:58 dominik@dell: ~/Code/openPMD/openPMD-example-datasets/example-thetaMode/hdf5 $ h5ls -r data00000100.h5 
/                        Group
/data                    Group
/data/100                Group
/data/100/fields         Group
/data/100/fields/B       Group
/data/100/fields/B/r     Dataset {3, 51, 201}
/data/100/fields/B/t     Dataset {3, 51, 201}
/data/100/fields/B/z     Dataset {3, 51, 201}
/data/100/fields/E       Group
/data/100/fields/E/r     Dataset {3, 51, 201}
/data/100/fields/E/t     Dataset {3, 51, 201}
/data/100/fields/E/z     Dataset {3, 51, 201}
/data/100/fields/J       Group
/data/100/fields/J/r     Dataset {3, 51, 201}
/data/100/fields/J/t     Dataset {3, 51, 201}
/data/100/fields/J/z     Dataset {3, 51, 201}
/data/100/fields/rho     Dataset {3, 51, 201}
```
B being the magnetic field and J the electric current, which can be neglected in electrostatic simulations. I think we should probably handle those as well - electrostatics are a fraction of all possible simulations. Do you think it would be much of a hassle to add them?

Reply:
A few tests for any of those datasets would also be very nice, of course.

Reply:
Sure! I'll work on them today.